File: src/common/py/api/component_events.py
import dataclasses

from .version import API_PROTOCOL_VERSION
from ..core.messaging import Event, Message
from ..core.messaging.builders import MessageBuilder, EventComposer
from ..utils import UnitID


@Message.define("event/component/information")
class ComponentInformationEvent(Event):
    """
    Contains information about a connected component; this is automatically sent whenever a connection is established (bilaterally).

    Attributes:
        comp_id: The component ID.
        comp_name: The component name.
        comp_version: The component version.
        api_protocol: The API protocol version.
    """

    comp_id: UnitID = dataclasses.field(default_factory=UnitID)

    comp_name: str = ""
    comp_version: str = ""

    api_protocol: str = str(API_PROTOCOL_VERSION)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        comp_id: UnitID,
        comp_name: str,
        comp_version: str,
        chain: Message | None = None,
    ) -> EventComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_event(
            ComponentInformationEvent,
            chain,
            comp_id=comp_id,
            comp_name=comp_name,
            comp_version=comp_version,
        )

File: src/common/web/api/ComponentEvents.ts
import { Type } from "class-transformer";

import { EventComposer } from "../core/messaging/composers/EventComposer";
import { MessageBuilder } from "../core/messaging/composers/MessageBuilder";
import { Event } from "../core/messaging/Event";
import { Message } from "../core/messaging/Message";
import { UnitID } from "../utils/UnitID";
import { API_PROTOCOL_VERSION } from "./Version";

/**
 * Layout of component information objects.
 */
export interface ComponentInformation {
    id: UnitID;
    name: string;
    version: string;
}

/**
 * Contains information about a connected component; this is automatically sent whenever a connection is established (bilaterally).
 *
 * @param comp_id - The component ID.
 * @param comp_name - The component name.
 * @param comp_version - The component version.
 * @param api_protocol - The API protocol version.
 */
@Message.define("event/component/information")
export class ComponentInformationEvent extends Event {
    // @ts-ignore
    @Type(() => UnitID)
    public readonly comp_id: UnitID = new UnitID("", "");

    public readonly comp_name: string = "";
    public readonly comp_version: string = "";

    public readonly api_protocol: string = String(API_PROTOCOL_VERSION);

    /**
     * The component information bundled as an object.
     *
     * @returns - The component information stored in a ``ComponentInformation`` object.
     */
    public componentInformation(): ComponentInformation {
        let info = {} as ComponentInformation;
        info.id = this.comp_id;
        info.name = this.comp_name;
        info.version = this.comp_version;
        return info;
    }

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, compID: UnitID, compName: string, compVersion: string,
                        chain: Message | null = null): EventComposer<ComponentInformationEvent> {
        return messageBuilder.buildEvent(ComponentInformationEvent,
            {
                comp_id: compID,
                comp_name: compName,
                comp_version: compVersion
            }, chain);
    }
}
